include/seqan3/io/alignment_file/record.hpp

```diff
@@ -0,0 +1,415 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2020, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2020, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::sam_record.
+ * \author Marcel Ehrhardt <marcel.ehrhardt AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/io/record.hpp>
+
+namespace seqan3
+{
+/*!\brief The record type of seqan3::alignment_file_input.
+ * \ingroup alignment_file
+ * \implements seqan3::detail::record_like
+ * \tparam field_types The types of the fields in this record as a seqan3::type_list.
+ * \tparam field_ids A seqan3::fields type with seqan3::field IDs corresponding to field_types.
+ */
+template <typename field_types, typename field_ids>
+class sam_record : public record<field_types, field_ids>
+{
+    //!\brief The base class.
+    using base_t = record<field_types, field_ids>;
+
+    //!\brief The underlying std::tuple class.
+    using tuple_base_t = typename base_t::base_type;
+
+    //!\copydoc seqan3::record::field_constant
+    template <field f>
+    using field_constant = typename base_t::template field_constant<f>;
+
+    using base_t::get_impl;
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    sam_record() = default; //!< Defaulted.
+    sam_record(sam_record const &) = default; //!< Defaulted.
+    sam_record & operator=(sam_record const &) = default; //!< Defaulted.
+    sam_record(sam_record &&) = default; //!< Defaulted.
+    sam_record & operator=(sam_record &&) = default; //!< Defaulted.
+    ~sam_record() = default; //!< Defaulted.
+
+    //!\brief Inherit std::tuple's and seqan3::record constructors.
+    using base_t::base_t;
+    //!\}
+
+    //!\brief The identifier, usually a string.
+    decltype(auto) id() &&
```

Comment:
Could you please explain why there are 4 different assignments for each field? The slight differences in syntax are not clear to me

Reply:
Yes, we want to mimic the way `std::get` behaves, and it will return 

* `id_type &&` for `std::get<id>(sam_record &&)`
* `id_type &` for `std::get<id>(sam_record &)`
* `id_type const &&` for `std::get<id>(sam_record const &&)`
* `id_type const &` for `std::get<id>(sam_record const &)`

We unfortunately need 4 overloads for this depending on whether the input has the type modifier `&&`, `&`, `const &&`, `const &`. The return type, here `decltype(auto)`, has the corresponding type modifier within it.

This is a special case, and this is not the norm, so you should not write 4 accessors functions if you write code yourself. In most cases it is fine just to copy the data without having a [ref/const-qualified function](https://en.cppreference.com/w/cpp/language/member_functions#ref-qualified_member_functions).

Reply:
Thanks for the explanation!